mlflow/openai/api_request_parallel_processor.py

```diff
@@ -71,279 +59,55 @@ def increment_num_rate_limit_errors(self):
-                    num_tokens += len(encoding.encode(value))
-                    if key == "name":  # if there's a name, the role is omitted
-                        num_tokens -= 1  # role is always required and always 1 token
-            num_tokens += 2  # every reply is primed with <im_start>assistant
-            return num_tokens + completion_tokens
-        # normal completions
-        else:
-            prompt = request_json["prompt"]
-            if isinstance(prompt, str):  # single prompt
-                prompt_tokens = len(encoding.encode(prompt))
-                return prompt_tokens + completion_tokens
-            elif isinstance(prompt, list):  # multiple prompts
-                prompt_tokens = sum([len(encoding.encode(p)) for p in prompt])
-                return prompt_tokens + completion_tokens * len(prompt)
-            else:
-                raise TypeError(
-                    "Expecting either string or list of strings for 'prompt' field in completion "
-                    "request"
-                )
-    # if embeddings request, tokens = input tokens
-    elif "embeddings" in request_url:
-        inp = request_json["input"]
-        if isinstance(inp, str):  # single input
-            return len(encoding.encode(inp))
-        elif isinstance(inp, list):  # multiple inputs
-            return sum([len(encoding.encode(i)) for i in inp])
-        else:
-            raise TypeError(
-                'Expecting either string or list of strings for "inputs" field in embedding request'
-            )
-    # more logic needed to support other API calls (e.g., edits, inserts, DALL-E)
-    else:
-        raise NotImplementedError(f'Support for "{request_url}" not implemented in this script')
+    import openai
+
+    status_tracker.start_task()
+    try:
+        result = task()
+        _logger.debug(f"Request #{index} succeeded")
+        status_tracker.complete_task(success=True)
+        results.append((index, result))
+    except openai.RateLimitError as e:
+        status_tracker.complete_task(success=False)
+        _logger.debug(f"Request #{index} failed with {e}")
+        status_tracker.increment_num_rate_limit_errors()
+        status_tracker.error = mlflow.MlflowException(
+            f"Request #{index} failed with rate limit: {e}."
+        )
+    except Exception as e:
+        status_tracker.complete_task(success=False)
+        _logger.debug(f"Request #{index} failed with {e}")
+        status_tracker.error = mlflow.MlflowException(f"Request #{index} failed with {e}")
 
 
 def process_api_requests(
-    requests: list[dict[str, any]],
-    request_url: str,
-    api_token: _OAITokenHolder,
-    # Reference: https://platform.openai.com/docs/guides/rate-limits/overview
-    max_requests_per_minute: float = 3_500,
-    max_tokens_per_minute: float = 90_000,
-    token_encoding_name: str = "cl100k_base",
-    max_attempts: int = 5,
+    requests: list[Callable],
```

Comment:
```suggestion
    requests: list[Callable[<arg type>, <return type>]],
```

Let's add argument type and return type.